Suggestion: Add overload for RegisterInContainer that pulls the DbContext from the service provider

Shane32 opened the pull request:
As it is now, it is required to pass an instance of the DbContext to the RegisterInContainer.  I suggest that the DbContext be pulled directly from the service provider instead.  The overload accepts functions to create a IModel based on the service provider, and has a default implementation built-in.  The functionality is demonstrated in SampleWeb, by commenting out the parameter that creates an instance of the database.  Also, for people that want to register the required services manually, ArgumentGraphs.RegisterInContainer was not in a public class (although the RegisterInContainer member was public), so I set the namespace and marked the class public.  It appeared to be an oversight based on the other code.

These changes should not conflict with my other pull requests.

* I am a patron under name of Zbox

SimonCropp commented on src/GraphQL.EntityFramework/Where/Graphs/ArgumentGraphs.cs:
```diff
@@ -4,32 +4,35 @@
 using GraphQL.Utilities;
 
-static class ArgumentGraphs
+namespace GraphQL.EntityFramework
```
can u remove the redundant namespace

SimonCropp commented:
can u add some doco for this

Shane32 commented:
The only changes are:
1. Adding the namespace. It was missing before so it’s not redundant
2. Add the word “public” before “static class ArgumentGraphs”

All the other changes that github shows is only because the contents are tabbed over. No actual changes. 

I’d be happy to add documentation. What type of documentation would you like me to add?  Just add comments throughout this class?

Shane32 commented:
Sorry that was for the first commit. I can add xml comments for the additional method. Probably tomorrow.

Shane32 commented:
I've updated the documentation to reflect the simpler method of registering the services.  Also, one of the old RegisterInContainer methods registered the connection types, while the other old RegisterInContainer method did not.  Based on the documentation, I removed it from the RegisterInContainer method that still had it in there.  This is a breaking change.  There are two specific commits if you wish to revert this behavior - 6143c9b and ef142ce

Shane32 commented:
And c8ac6bf ...

SimonCropp commented:
@Shane32 i think i will revert this. it is too problematic extracting imodel from the container when thedbcontext should be a scopes service. ie we are forcing a singleton to rely on a scoped service

Shane32 commented:
The DbContext is scoped no matter how you cut it.  The problem is, if you are required to pass it in to the constructor of the register function, then you cannot take advantage of dependency injection to construct the dbcontext to begin with, as the service provider is not available (and even if it were, all the services may not yet have been registered, and dependency injection is designed to be registered in any order).  So then it requires a manual instance of the dbcontext.  This brings us back to the exact problem with the sample code as it previously existed -- an ugly hack because the service provider is unavailable to provide the connection string.

Keep in mind that the singleton is still a singleton.  It is created once, upon the first use of the service, and creates the dbcontext in the intended fashion - from the service provider - and disposes of it as soon as it captures the imodel.  The singleton does not hold a reference to the scoped dbcontext, and does not, for the purposes of dependency injection, rely on a scoped service.  Functionally this is identical to the prior code -- capturing an IModel from a dbcontext and immediately disposing of the dbcontext.  The difference is the time at which it is captured -- in this case, upon first use, rather than during setup of the service provider.  (Creating upon first use is really how it should be done anyway.)

In my particular scenario, the dbcontext OnModelCreating function is run through dependency injection, so that the different data models (tables) can be developed independently and added to the dbcontext without adding or editing code in the dbcontext.  So creating a dbcontext during service registration is impossible -- or I'd have to manually register each database table with the dbcontext, entirely nullifying the point of dependency injection to begin with.

I'm a bit confused about your comment that this is "problematic".  Many services, such as the dbcontext itself, rely on the service provider to configure themselves.  This one is no different.  It actually makes the service registration a lot easier, as there is no need to hack a dbcontext for the registration.

Personally, I would suggest that the GlobalFilters pull from the service provider by default as well.  However, I am not using them, and don't know how they would be used.  With the code snippet I provided, someone could easily pass a function like `(serviceProvider) => serviceProvider.GetRequiredService<GlobalFilters>()` to the registration function to accomplish this, although the default is currently just an empty GlobalFilters class, as it was in the prior version of this code.  I'd also personally remove the extra Func<> parameters and always automatically obtain the IModel and GlobalFilters from the service provider, letting people manually register the services if they wish to override the defaults.  Or perhaps a configuration Action<> could be used like is typical for dependency injection.  These suggestions I did not implement in my pull request, as I was not sure you would agree.  One final suggestion, which I have not yet explored, is to pull the dbcontext from dependency injection within the graphs.  This might be tricky to pull off, but then it would not require embedding the dbcontext within the usercontext of the graph or a GetDbContextFromUserContext function.

If you disagree with this pull request, I'd suggest that you at least make the ArgumentGraphs class public ( 7eaaec8 ) , so anyone can manually register the required classes rather than relying on the RegisterInContainer function when it may not work for their needs.
